### ISSUE TITLE
Provide for searching by public or long name in catalog

### ISSUE THREAD

**Issue by MikeNeilson** [USACE/cwms-data-api#288]
While typing up a searching for data guide I realized a lot of our location names are abbreviations that a lot of people may not be familiar with.

The catalog should allow searching by public or long name.

**Comment by adamkorynta** [comment]
Location kind would be helpful too given hec-cwms-data-access provides a retrieval method by kind: https://bitbucket.hecdev.net/projects/CWMS/repos/hec-cwms-data-access/browse/hec-db-jdbc/src/main/java/wcds/dbi/oracle/cwms/CwmsLocationJdbcDao.java#446

**Comment by rma-rripken** [comment]
I recently worked on the Location Catalog.  There are currently 4 optional regex filters.   The method signature is getting a little out of hand.   I suggest we change the Location Dao interface so that it has a single Map<String, String> of regex filters where the Key is the name of the field and the value is the regex.   Then we wouldn't have to change the Dao each time we want to add another potential regex filter.    In the Dao we can loop over all the filters entries and add a condition for each filter that was specified.   It would be fairly easy to add a filter for every String field in LocationCatalogEntry if we wanted.

**Comment by rma-rripken** [comment]
Actually, maybe we should add filters for the boolean and double fields as well.   For example, being able to query for locations inside a geographic box could be very useful and would be easy to implement.

**Comment by MikeNeilson** [comment]
Agreed. And I'm quite sure that particular feature (geographic box) will be asked for soon.

**Comment by rma-rripken** [comment]
The Location catalog was recently changed to add CatalogRequestParameters and that means it would now be fairly easy to add filters for public-name or long-name on location.    I envision similar changes for the timeseries catalog but that has not been done yet.

**Comment by MikeNeilson** [comment]
If more people start asking we'll knock the priority up.

**Comment by david-a-ries** [comment]
Hi all,

I'd like to put in a request for adding filtering to the location catalog to include filtering by "public-name" in CDA.  This has been a topic of conversation in NWDP (primarily for the web migration) as a way to make it easier to return timeseries without having to know the 3/4 letter code for the time series.

There are a few examples in our database where there 3/4 letter codes don't match up with the "public-name" well.  For example one of our projects is often referred to as "Mason Dam" with a "public-name" of "Mason Dam&Phillips Lk nr Baker" where the 3 letter code for the dam is based on the name of the lake and is "PHL."  It may be difficult for someone to know that the project often referred to as "Mason Dam" has a 3 letter code of "PHL."  

Even in situations where the naming isn't as different as this example, it's harder for the end user to know exactly how to find what location their actually looking for.  It would also allow for returning multiple locations that are associated with certain projects/areas.  For example, searching for "Lower Granite" would be able to return the location id of the Dam, the fish ladder temperatures, temperature strings, tailwater radar stations, and the TDG stations associated with the project that aren't directly associated with the dam or under the same Location ID.

Overall, having a "public-name" filter in the Locations catalog endpoint would make it much easier to return possible locations/timeseries that the user is interested in and would be a valuable tool to have for a dataquery tool.  NWDP's current internal and public version of dataquery relies on a configuration file that matches the public names to the location IDs.  It would be possible to do something like this and store that config file in the Blob endpoint; however, that file would have to be maintained and it would be much easier to have the search by "public-name" built into the Location Catalog endpoint.

**Comment by rma-rripken** [comment]
Anyone working on this should read  https://github.com/USACE/cwms-data-api/discussions/935   There are other columns that users would also like to search.

**Comment by MikeNeilson** [comment]
@JeremyDKellett was also looking into some full text search indexes, which should give "any column" fairly easily.

Unfortunately I don't know what the timeless is on that right now. We should probably open an issue over at https://github.com/hydrologicengineeringcenter/cwms-database to track it.

**Comment by LaurenAllin** [comment]
Similar to NWDP, MVP would like the option to search the catalog on public-name and location id/name. This issue was closed, but looking at CDA docs on Swagger I am not sure how to complete the CDA call to search on public-name or long-name, and only see an option to search on name. Was this feature/ability added to CDA?

**Comment by MikeNeilson** [comment]
> Similar to NWDP, MVP would like the option to search the catalog on public-name and location id/name. This issue was closed, but looking at CDA docs on Swagger I am not sure how to complete the CDA call to search on public-name or long-name, and only see an option to search on name. Was this feature/ability added to CDA?

The change hasn't made it's way to prod, or even test, yet. If you go to the dev URL in the swagger-ui for "Catalog" there's a `search-text` parent that will be at the bottom of the list of parameters. That searches everything about a location, including public name, for the given text.

**Comment by LaurenAllin** [comment]
Thanks for clarifying! I had a quick test and it looks to be working as I would expect it to. Groundwork Water is hoping to use this endpoint for a Search Bar component whenever it is available in CDA prod.
